Add missing SSL context options

trowski opened the pull request:
Adds ~~`allow_self_signed`, `disable_compression`,~~ and `peer_fingerprint` options to `ClientTLSContext` that are currently missing.

Also adds `passphrase` to `Certificate`.

Edit: Removed `disable_compression` and `allow_self_signed`.

trowski committed:
Add missing SSL context options

trowski committed:
Consistency

Bilge commented on src/ClientTlsContext.php:
```diff
@@ -24,6 +24,12 @@ final class ClientTlsContext
 
+    private bool $allowSelfSigned = false;
+
+    private bool $disableCompression = true;
```
Why is this enabled by default?

trowski replied:
You mean disabled by default? Because that is PHP's default.

Bilge replied:
I meant why is this setting enabled. Perhaps it should just be called _compression_ instead of _disable compression_. Anyway, how do we know what the PHP default is? It's not [documented](https://www.php.net/manual/en/context.ssl.php)...

trowski replied:
Naming is terrible yes, but it's mirroring the context option naming.

The default was changed in PHP 5.6: https://www.php.net/manual/en/migration56.openssl.php

trowski replied:
@kelunik raised a good point in chat – compression is removed in future TLS version anyway, so there's no reason to have an option for it.

trowski committed:
Drop disable_compression

Insecure and has been removed from future TLS versions.

trowski committed:
Also remove allow_self_signed

trowski committed:
Test peer fingerprint verification

trowski commented:
I also now removed `allow_self_signed` since `verify_peer` _does not_ need to be enabled for `peer_fingerprint` to be validated. I added a note to the README about self-signed certs explaining the missing option to allow self-signed certs and pointing users toward using peer fingerprint instead.

kelunik commented on src/ClientTlsContext.php:
```diff
@@ -366,6 +368,46 @@ public function getCertificate(): ?Certificate
+        return $clone;
+    }
+
+    public function getPeerFingerprint(): ?array
```
```suggestion
    public function getPeerFingerprints(): ?array
```

trowski committed:
Pluralize fingerprint

kelunik commented on src/ClientTlsContext.php:
```diff
@@ -366,6 +368,45 @@ public function getCertificate(): ?Certificate
         return $this->certificate;
     }
 
+    public function withPeerFingerprints(string|array $fingerprint): self
```
Should we make this two separate methods, `withPeerFingerprint(string)` and `withPeerFingerprints(array)`?

trowski replied:
Sure, that seems a bit cleaner. Done!

trowski committed:
Separate methods for string and array fingerprints
